util/topo/provider_cached.go

```diff
@@ -0,0 +1,123 @@
+// Copyright 2021 PingCAP, Inc. Licensed under Apache-2.0.
+
+package topo
+
+import (
+	"context"
+	"runtime"
+	"time"
+
+	"github.com/ReneKroon/ttlcache/v2"
+)
+
+// CachedTopology provides topology over an underlying topology provider with a TTL cache.
+// This struct is concurrent-safe.
+type CachedTopology struct {
+	p     TopologyProvider
+	cache *ttlcache.Cache
+}
+
+var _ TopologyProvider = (*CachedTopology)(nil)
+
+func NewCachedTopology(p TopologyProvider, ttl time.Duration) *CachedTopology {
+	cache := ttlcache.NewCache()
+	cache.SkipTTLExtensionOnHit(true)
+	_ = cache.SetTTL(ttl)
+
+	ct := &CachedTopology{
+		p:     p,
+		cache: cache,
+	}
+	// Destroy the internal cache goroutine when no one is referencing this struct anymore.
+	runtime.SetFinalizer(ct, (*CachedTopology).finalize)
+	return ct
+}
+
+func (c *CachedTopology) getOrFillCache(key string, backSource func() (interface{}, error)) (interface{}, error) {
+	if data, err := c.cache.Get(key); err == nil {
+		return data, nil
+	}
+	// TODO: use singleflight.
+	src, err := backSource()
+	if err != nil {
+		// Error is never cached.
+		return nil, err
+	}
+	_ = c.cache.Set(key, src)
+	runtime.KeepAlive(c)
```

Comment:
Ok...This is too elaborate. Then we also need to be careful about using `c.cache` outside of `getOrFillCache` in the future.

Reply:
Probably won't use `c.cache` anymore.

Reply:
Cool, knowledge++

Reply:
Yeah. Probably I should add a comment: Don't touch this unless you are professional about `runtime.SetFinalizer` behavior.. LOL